edi/models/edi_exchange_record.py

```diff
@@ -369,10 +372,16 @@ def check_access_rule(self, operation):
         if self.env.is_superuser():
             return
         for exchange_record in self:
-            if not exchange_record.record:
+            sudo_record = exchange_record.sudo()
+            if not sudo_record.model or not sudo_record.res_id:
                 continue
-            if hasattr(exchange_record.record, "get_edi_access"):
-                check_operation = exchange_record.record.get_edi_access(
+            record = (
```

Comment:
same, would be better to do it for the whole loop IMO

Reply:
The problem is that some records could have the edi_access and others not, so it is hard to remove the logic from here :thinking:

Reply:
I'd do something like this (completely untested :stuck_out_tongue: :


```py
        default_checker = self.env[
            "edi.exchange.consumer.mixin"
        ].get_edi_access
        exchange_records = self.sudo().filtered(lambda x: x.model and x.res_id)
        by_model_rec_ids = defaultdict(set)
        by_model_checker = {}
        for exc_rec in exchange_records:
            by_model_rec_ids[exc_rec.model].add(exc_rec.id)
            checker = getattr(self.env[exc_rec.model], "get_edi_access", default_checker)

        for model, rec_ids in by_model_rec_ids.items():
            records = self.env[model].browse(rec_ids).with_user(self._uid)
            checker = by_model_checker[model]
            for record in records:
                check_operation = checker([record.id], operation, model_name=record._name)
                record.check_access_rights(check_operation)
                record.check_access_rule(check_operation)
```

sounds good?

Reply:
BTW would you mind adding a test for permissions? Otherwise please add a TODO

Reply:
Tests where already there :wink:  https://github.com/OCA/edi/blob/13.0/edi/tests/test_security.py

I adapted your code, as it seems cleaner :smile: